samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java

```diff
@@ -129,7 +129,7 @@ public void close() {
    */
   OperatorImpl createAndRegisterOperatorImpl(OperatorSpec prevOperatorSpec, OperatorSpec operatorSpec,
       Config config, TaskContext context) {
-    if (!operatorImpls.containsKey(operatorSpec) || operatorSpec instanceof JoinOperatorSpec) {
+    if (!operatorImpls.containsKey(operatorSpec.getOpName()) || operatorSpec instanceof JoinOperatorSpec) {
```

Comment:
This relies on the guarantee that getOpName() returns the unique ID for an operatorSpec in a user defined operator graph. Instead, using the operatorSpec object reference directly as key does not require that. Curious about what's the advantage here using String as the key?